gef-ui/server/controllers/unissued-facilities/facility-end-date/index.ts

```diff
@@ -0,0 +1,217 @@
+import * as api from '../../../services/api';
+import { validateAndParseFacilityEndDate } from './validation';
+import { asLoggedInUserSession } from '../../../utils/express-session';
+import { FacilityEndDateViewModel } from '../../../types/view-models/facility-end-date-view-model';
+import { ValidationError } from '../../../types/validation-error';
+
+type FacilityEndDateParams = { dealId: string; facilityId: string };
+type FacilityEndDatePostBody = { 'facility-end-date-day': string; 'facility-end-date-month': string; 'facility-end-date-year': string };
+
+type GetFacilityEndDateRequest = CustomExpressRequest<{ params: FacilityEndDateParams; query: { change: string } }>;
+type PostFacilityEndDateRequest = CustomExpressRequest<{
+  reqBody: FacilityEndDatePostBody;
+  params: FacilityEndDateParams;
+  query: { saveAndReturn: string };
+}>;
+
+const getFacilityEndDateViewModel = async (req: GetFacilityEndDateRequest, previousPage: string): Promise<FacilityEndDateViewModel | null> => {
+  const {
+    params: { dealId, facilityId },
+    session: { userToken },
+  } = req;
+
+  const { details: facility } = (await api.getFacility({ facilityId, userToken })) as { details: Record<string, unknown> };
+  const deal = (await api.getApplication({ dealId, userToken })) as Record<string, unknown> & { version?: number };
+
+  if (!isFacilityEndDateEnabledOnGefVersion(parseDealVersion(deal.version)) || !facility.isUsingFacilityEndDate) {
+    return null;
+  }
+
+  const facilityEndDateViewModel: FacilityEndDateViewModel = {
+    dealId,
+    facilityId,
+    previousPage,
+  };
+
+  if ('facilityEndDate' in facility && typeof facility.facilityEndDate === 'string') {
+    const facilityEndDate = parseISO(facility.facilityEndDate);
+    facilityEndDateViewModel.facilityEndDate = {
+      day: String(getDate(facilityEndDate)),
+      month: String(getMonth(facilityEndDate) + 1),
+      year: String(getYear(facilityEndDate)),
+    };
+  }
+
+  return facilityEndDateViewModel;
+};
+
+export const getFacilityEndDateFromUnissuedFacilitiesPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/about`;
+
+  try {
+    const facilityEndDateViewModel = await getFacilityEndDateViewModel(req, previousPage);
+
+    if (!facilityEndDateViewModel) {
+      return res.redirect(previousPage);
+    }
+
+    return res.render('partials/facility-end-date.njk', facilityEndDateViewModel);
+  } catch (error) {
+    console.error(error);
+    return res.render('partials/problem-with-service.njk');
+  }
+};
+
+export const getFacilityEndDateFromApplicationPreviewPage = async (req: GetFacilityEndDateRequest, res: Response) => {
+  const {
+    params: { dealId, facilityId },
+  } = req;
+
+  const previousPage = `/gef/application-details/${dealId}/unissued-facilities/${facilityId}/change`;
+
+  try {
```

Comment:
were we going to commonise this one and the above one?

Reply:
Dont think i follow - the `getFacilityEndDateViewModel` is the commonised bit